ethdb/memorydb/memorydb.go

```diff
@@ -279,25 +280,24 @@ func (b *batch) Replay(w ethdb.KeyValueWriter) error {
 // value store. Internally it is a deep copy of the entire iterated state,
 // sorted by keys.
 type iterator struct {
-	inited bool
+	index  int
 	keys   []string
 	values [][]byte
 }
 
 // Next moves the iterator to the next key/value pair. It returns whether the
 // iterator is exhausted.
 func (it *iterator) Next() bool {
-	// If the iterator was not yet initialized, do it now
-	if !it.inited {
-		it.inited = true
-		return len(it.keys) > 0
+	// Short circuit if iterator has nothing to iterate.
+	if len(it.keys) == 0 {
+		return false
 	}
-	// Iterator already initialize, advance it
-	if len(it.keys) > 0 {
-		it.keys = it.keys[1:]
-		it.values = it.values[1:]
+	// Short circuit if iterator is already exhausted in the forward direction.
+	if it.index == len(it.keys) {
+		return false
 	}
-	return len(it.keys) > 0
+	it.index += 1
+	return it.index != len(it.keys)
```

Comment:
Wouldn't this be simple as:

```go
func (it *iterator) Next() bool {
    if it.index + 1 >= len(it.key) {
        return false
    }
    it.index++
    return true
}
```

? Equality check would also work, but `>=` seems correcter to digest.

Reply:
yup, looks better, I will apply

Reply:
Ah, after applying the change, some tests are broken. So essentially the behavior is changed.

Previously if we are at the last element, `Next` call will shift the position to `EOI`. While with your change, the iterator position won't be changed in this case. So it might have some side effects, for example we can still read key or value from an exhausted iterator.

Reply:
Changing the last check is simply easier on the eye, it's clearer that we're looking for "smaller" indices, not unequal ones. 

I'd probably also change the `==` one to `>=` for the same reasons.